external/fv3fit/fv3fit/keras/_models/models.py

```diff
@@ -308,11 +316,23 @@ def _fit_loop(
                 )
 
     def predict(self, X: xr.Dataset) -> xr.Dataset:
-        sample_coord = X[self.sample_dim_name]
-        ds_pred = self.y_packer.to_dataset(
-            self.predict_array(self.X_packer.to_array(X))
+        # Takes unstacked data, stacks into sample dimension before
+        # keras model prediction, and returns unstacked prediction
+        # ensure dimension order is the same
+        inputs_ = safe.get_variables(X, self.input_variables)
+        inputs_stacked = stack_non_vertical(inputs_)
+        sample_coord = inputs_stacked[self.sample_dim_name]
+        ds_pred_stacked = self.y_packer.to_dataset(
+            self.predict_array(self.X_packer.to_array(inputs_stacked))
         )
-        return ds_pred.assign_coords({self.sample_dim_name: sample_coord})
+        ds_pred = ds_pred_stacked.assign_coords(
+            {self.sample_dim_name: sample_coord}
+        ).unstack(self.sample_dim_name)
+        ds_pred = match_prediction_to_input_coords(X, ds_pred)
+        dim_order = [
+            dim for dim in infer_dimension_order(inputs_) if dim in ds_pred.dims
+        ]
```

Comment:
It feels like there are a lot of low level details here. Is it possible to split this into three parts? Stacking the sample dim, prediction on stacked data, and unstacking the sample dim? Ideally these would each be function/method calls that self-document what they're doing, instead of needing the 3-line comment above.

If it's possible to put the stacking/unstacking into a helper function that's used the same by _random_forest.py, testing.py, and here, that would be ideal.

Reply:
I moved the "predict on stacked data" parts that were unique to each model into their own method, which is called between the common 1st/3rd parts in `.predict`. It seemed more clear to leave most of the 1st/3rd parts outside of functions but lmk if you disagree. I consolidated the dimension ordering into the `match_prediction_to_input_coords` helper function to tidy up the last part.